File: src/main/java/tmi/recipe/parser/ThermalGeneratorParser.java
package tmi.recipe.parser;

import arc.math.Mathf;
import arc.struct.Seq;
import mindustry.Vars;
import mindustry.content.Blocks;
import mindustry.world.Block;
import mindustry.world.blocks.environment.Floor;
import mindustry.world.blocks.power.ThermalGenerator;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeItemStack;
import tmi.recipe.RecipeType;
import tmi.recipe.types.PowerMark;

public class ThermalGeneratorParser extends ConsumerParser<ThermalGenerator>{
  {excludes.add(GeneratorParser.class);}

  @Override
  public boolean isTarget(Block content) {
    return content instanceof ThermalGenerator;
  }

  @Override
  public Seq<Recipe> parse(ThermalGenerator content) {
    Recipe res = new Recipe(RecipeType.generator);
    res.block = content;

    registerCons(res, content.consumers);

    res.addProductionPresec(PowerMark.INSTANCE, content.powerProduction);

    for (Block block : Vars.content.blocks()) {
      if (block.attributes.get(content.attribute) <= 0) continue;

      float eff = content.displayEfficiencyScale*content.size*content.size*block.attributes.get(content.attribute);
      if (eff <= content.minEfficiency) continue;
      res.addMaterial(block, content.size*content.size)
          .setEfficiency(eff)
          .setAttribute()
          .setFormat(f -> "[#98ffa9]" + Mathf.round(eff*100) + "%");
    }

    if (content.outputLiquid != null) res.addProductionPresec(content.outputLiquid.liquid, content.outputLiquid.amount);

    return Seq.with(res);
  }
}

File: src/main/kotlin/tmi/recipe/parser/ThermalGeneratorParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.Vars
import mindustry.world.Block
import mindustry.world.blocks.power.ThermalGenerator
import tmi.recipe.Recipe
import tmi.recipe.RecipeItemGroup
import tmi.recipe.RecipeParser
import tmi.recipe.types.RecipeItemType
import tmi.recipe.RecipeType
import tmi.recipe.types.PowerMark

open class ThermalGeneratorParser : ConsumerParser<ThermalGenerator>() {
  override val excludes: Seq<Class<out RecipeParser<*>>> = Seq.with(GeneratorParser::class.java)

  override fun isTarget(content: Block): Boolean {
    return content is ThermalGenerator
  }

  override fun parse(content: ThermalGenerator): Seq<Recipe> {
    val res = Recipe(
      recipeType = RecipeType.generator,
      ownerBlock = content.getWrap()
    ).setBaseEff(0f)

    registerCons(res, *content.consumers)

    res.addProductionPersec(PowerMark, content.powerProduction/content.displayEfficiencyScale)
      .setType(RecipeItemType.POWER)

    val attrGroup = RecipeItemGroup()
    for (block in Vars.content.blocks()) {
      if (content.attribute == null || block.attributes[content.attribute] <= 0) continue

      val eff = content.size*content.size*block.attributes[content.attribute]
      if (eff <= content.minEfficiency) continue
      res.addMaterial(block.getWrap(), (content.size*content.size) as Number)
        .setEfficiency(eff*content.displayEfficiencyScale)
        .setType(RecipeItemType.ATTRIBUTE)
        .efficiencyFormat(eff*content.displayEfficiencyScale)
        .setGroup(attrGroup)
    }

    if (content.outputLiquid != null) res.addProductionPersec(
      content.outputLiquid.liquid.getWrap(),
      content.outputLiquid.amount/content.displayEfficiencyScale
    )

    return Seq.with(res)
  }
}
